pkg/tools/etcd_helper.go

```diff
@@ -38,6 +39,8 @@ type EtcdHelper struct {
 	Codec  runtime.Codec
 	// optional, no atomic operations can be performed without this interface
 	Versioner EtcdVersioner
+	cache     map[uint64]reflect.Value
```

Comment:
Please add a comment explaining this cache... including why you chose reflect.Value to store instead of runtime.Object, that it depends on etcd's indexes being globally unique per object (so if we make multiple etcd clusters, we'd need to include object key as part of this key), and that it's limited to N entries.